### PR TITLE
Add `vue/no-unregistered-components` rule

### PR DESCRIPTION
**Pull request by jesusgn90**
Closes https://github.com/vuejs/eslint-plugin-vue/issues/1112

_Inspired by [rules/no-unused-components](https://eslint.vuejs.org/rules/no-unused-components.html) so thanks to Michał Sajnóg._

What are your thoughts @ota-meshi? let me know if this is missing something or you have some concerns, thanks!

### CODE REVIEW

File: lib/rules/no-unregistered-components.js [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+// ------------------------------------------------------------------------------
+// Rule Definition
+// ------------------------------------------------------------------------------
+
+module.exports = {
+  meta: {
+    type: 'suggestion',
+    docs: {
+      description: 'disallow using components that are not registered',
+      categories: ['essential'],
```
Comment (ota-meshi) [inline lib/rules/no-unregistered-components.js:23]: The user must configure the globally registered components. So the category cannot be `'essential'`.
Please set `undefined`.

File: tests/lib/rules/no-unregistered-components.js [hunk 2]
```diff
@@ -0,0 +1,385 @@
...
+          data: () => ({
+            showComponent: false
+          }),
+          components: {
+            CustomComponent
+          }
+        }
+        </script>
+      `
+    }
```
Comment (ota-meshi) [inline tests/lib/rules/no-unregistered-components.js:235]: Could you add a test case to check if the built-in component is not reported?

https://vuejs.org/v2/api/#Built-In-Components

File: tests/lib/rules/no-unregistered-components.js [hunk 3]
```diff
@@ -0,0 +1,385 @@
...
+      filename: 'test.vue',
+      code: `
+        <template>
+          <CustomComponent />
+        </template>
+      `,
+      options: [
+        {
+          ignorePatterns: [
+            /custom(\-\w+)+/
```
Comment (ota-meshi) [inline tests/lib/rules/no-unregistered-components.js:46]: Please define all ignorePatterns as strings.

File: docs/rules/no-unregistered-components.md [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+sidebarDepth: 0
+title: vue/no-unregistered-components
+description: disallow using components that are not registered inside templates
+---
+# vue/no-unregistered-components
+> disallow using components that are not registered inside templates
+
+## :book: Rule Details
+
+This rule reports components that haven't been registered and are being used in the template.
```
Comment (ota-meshi) [inline docs/rules/no-unregistered-components.md:12]: Could you document that this rule cannot check globally registered components and components registered in mixins?

File: docs/rules/no-unregistered-components.md [hunk 5]
```diff
@@ -0,0 +1,132 @@
...
+  }]
+}
+```
+
+- `ignorePatterns` ... suppresses all errors if component name matches one or more patterns
+    default `[]`
+
+### `ignorePatterns: [/custom(\-\w+)+/]`
+
+<eslint-code-block :rules="{'vue/no-unregistered-components': ['error', { 'ignorePatterns': [/custom(\-\w+)+/] }]}">
```
Comment (ota-meshi) [inline docs/rules/no-unregistered-components.md:85]: IgnorePatterns should be defined as a string.

```suggestion
### `"ignorePatterns": ["/custom(\-\w+)+/"]`

<eslint-code-block :rules="{'vue/no-unregistered-components': ['error', { 'ignorePatterns': ['/custom(\-\w+)+/'] }]}">
```

File: docs/rules/no-unregistered-components.md [hunk 6]
```diff
@@ -0,0 +1,132 @@
...
+    components: {
+
+    },
+  }
+</script>
+```
+
+</eslint-code-block>
+
+<eslint-code-block :rules="{'vue/no-unregistered-components': ['error', { 'ignorePatterns': [/custom(\-\w+)+/] }]}">
```
Comment (ota-meshi) [inline docs/rules/no-unregistered-components.md:107]:
```suggestion
<eslint-code-block :rules="{'vue/no-unregistered-components': ['error', { 'ignorePatterns': ['/custom(\-\w+)+/'] }]}">
```

File: tests/lib/rules/no-unregistered-components.js [hunk 7]
```diff
@@ -0,0 +1,385 @@
...
+      code: `
+        <template>
+          <customComponent>
+            Some text
+          </customComponent>
+        </template>
+      `,
+      options: [
+        {
+          ignorePatterns: [
```
Comment (ota-meshi) [inline tests/lib/rules/no-unregistered-components.js:130]: Could you add test case to test multiple ignorePatterns?

File: tests/lib/rules/no-unregistered-components.js (hunk 2 above)
Comment (ota-meshi) [reply tests/lib/rules/no-unregistered-components.js:235]: Also, we are working on support for Vue3, so please do not report `<teleport>` and `<suspense>`.

File: lib/rules/no-unregistered-components.js [hunk 8]
```diff
@@ -0,0 +1,108 @@
...
+      "VAttribute[directive=false][key.name='is']" (node) {
+        usedComponentNodes.push({ node, name: node.value.value })
+      },
+      "VElement[name='template']" (rootNode) {
+        templateLocation = templateLocation || rootNode.loc.start
+      },
+      "VElement[name='template']:exit" (rootNode) {
+        if (
+          rootNode.loc.start !== templateLocation ||
+          utils.hasAttribute(rootNode, 'src')
```
Comment (ota-meshi) [inline lib/rules/no-unregistered-components.js:78]: What does this branch condition mean?

File: lib/rules/no-unregistered-components.js [hunk 9]
```diff
@@ -0,0 +1,108 @@
...
+          utils.isHtmlWellKnownElementName(node.rawName) ||
+          utils.isSvgWellKnownElementName(node.rawName) ||
+          node.rawName === 'component'
+        ) {
+          return
+        }
+
+        usedComponentNodes.push({ node, name: node.rawName })
+      },
+      "VAttribute[directive=true][key.name.name='bind'][key.argument.name='is']" (node) {
```
Comment (ota-meshi) [inline lib/rules/no-unregistered-components.js:58]: You can also specify an HTML element for `is`.

e.g.
 
```vue
<component is="div">
<component :is="'div'">
```

Could you change it to exclude these?

File: lib/rules/no-unregistered-components.js [hunk 10]
```diff
@@ -0,0 +1,108 @@
...
+            const kebabCaseName = casing.kebabCase(name)
+            if (ignorePatterns.find(pattern => {
+              const regExp = new RegExp(pattern)
+              return regExp.test(kebabCaseName) ||
+                regExp.test(casing.pascalCase(name)) ||
+                regExp.test(casing.camelCase(name)) ||
+                regExp.test(casing.snakeCase(name)) ||
+                regExp.test(name)
+            })) return false
+            return registeredComponentNames.indexOf(kebabCaseName) === -1
```
Comment (ota-meshi) [inline lib/rules/no-unregistered-components.js:94]: Components registered in Pascal Case can be used in kebab-case, but components registered in kebab-case cannot be used in Pascal Case.

Perhaps the next use would make this rule a false negative.

```vue
<template>
  <FooBar/>
</template>
<script>
export default {
  components: {
    'foo-bar': FooBarComponent
  }
}
</script>
```

File: lib/rules/no-unregistered-components.js (hunk 1 above)
Comment (jesusgn90) [reply lib/rules/no-unregistered-components.js]: You are right, I've changed it :+1:

File: tests/lib/rules/no-unregistered-components.js (hunk 2 above)
Comment (jesusgn90) [reply tests/lib/rules/no-unregistered-components.js:322]: I've added some extra logic for these cases and different tests to prove this.

File: tests/lib/rules/no-unregistered-components.js (hunk 3 above)
Comment (jesusgn90) [reply tests/lib/rules/no-unregistered-components.js]: This is now done and examples are updated :+1:

File: docs/rules/no-unregistered-components.md (hunk 4 above)
Comment (jesusgn90) [reply docs/rules/no-unregistered-components.md:12]: I've added a warning box telling this and explaining that `component`, `suspense` and `teleport` are exceptions for this.

File: docs/rules/no-unregistered-components.md (hunk 5 above)
Comment (jesusgn90) [reply docs/rules/no-unregistered-components.md]: This is now done and examples are updated :+1:

File: docs/rules/no-unregistered-components.md (hunk 6 above)
Comment (jesusgn90) [reply docs/rules/no-unregistered-components.md]: This is now done and examples are updated :+1:

File: tests/lib/rules/no-unregistered-components.js (hunk 7 above)
Comment (jesusgn90) [reply tests/lib/rules/no-unregistered-components.js:130]: I've added a test with multiple patterns :+1:

File: lib/rules/no-unregistered-components.js (hunk 9 above)
Comment (jesusgn90) [reply lib/rules/no-unregistered-components.js:80]: Now it's excluding uses cases like for any well known HTML element.

File: lib/rules/no-unregistered-components.js (hunk 10 above)
Comment (jesusgn90) [reply lib/rules/no-unregistered-components.js:117]: Added a test to prove this is working.

**Comment by jesusgn90** [comment]
@ota-meshi I think all the concerns are now fixed, tests are passing locally but the CI is blocking for `netlify` parts somehow and I don't know why, `node` CIs are passing though.

**Comment by jesusgn90** [comment]
Seems like `netlify` parts are now happy, maybe the `suspense` and `teleport` corrections did the trick

File: tests/lib/rules/no-unregistered-components.js (hunk 2 above)
Comment (ota-meshi) [reply tests/lib/rules/no-unregistered-components.js:322]: Could you add a test case to ensure that the following components are not reported?

- `<transition>`
- `<transition-group>`
- `<keep-alive>`

https://vuejs.org/v2/api/#Built-In-Components

File: lib/rules/no-unregistered-components.js (hunk 10 above)
Comment (ota-meshi) [reply lib/rules/no-unregistered-components.js:117]: Which test case is it? I couldn't find it.

File: lib/rules/no-unregistered-components.js (hunk 9 above)
Comment (ota-meshi) [reply lib/rules/no-unregistered-components.js:80]: Could you please add the following test cases?

```vue
<component is="div">
```

File: lib/rules/no-unregistered-components.js (hunk 8 above)
Comment (ota-meshi) [reply lib/rules/no-unregistered-components.js:101]: Perhaps it doesn't make sense so could you remove this conditional?

File: package.json [hunk 11]
```diff
@@ -5,7 +5,7 @@
   "main": "lib/index.js",
   "scripts": {
     "start": "npm run test:base -- --watch --growl",
-    "test:base": "mocha \"tests/lib/**/*.js\" --reporter dot",
+    "test:base": "mocha \"tests/lib/**/no-unregistered-components.js\" --reporter dot",
```
Comment (ota-meshi) [inline package.json:8]: Do not change the test script.

File: lib/rules/no-unregistered-components.js (hunk 10 above)
Comment (jesusgn90) [reply lib/rules/no-unregistered-components.js:117]: https://github.com/vuejs/eslint-plugin-vue/pull/1114/files#diff-1694a97f87118f4447a8f24732f687f0R292-R303

```js
// Valid
{
      filename: 'test.vue',
      code: `
        <template>
          <CustomComponent />
        </template>
        <script>
        export default {
          components: {
            'custom-component': InfoPrimaryWrapper
          }
        }
        </script>
      `
},
```

File: package.json (hunk 11 above)
Comment (jesusgn90) [reply package.json:8]: Yep, this was not intended

**Comment by jesusgn90** [comment]
Ok I think I'm almost there, just missing the `<component is="div">` edge case (atm supports only `<component :is="'div'">`, I'll re-request review once done

File: lib/rules/no-unregistered-components.js [hunk 12]
```diff
@@ -0,0 +1,134 @@
...
+          if (utils.isHtmlWellKnownElementName(node.value.expression.value)) return
+          usedComponentNodes.push({ node, name: node.value.expression.value })
+        }
+      },
+      "VAttribute[directive=false][key.name='is']" (node) {
+        if (utils.isHtmlWellKnownElementName(node.value.value)) return
+        usedComponentNodes.push({ node, name: node.value.value })
+      },
+      "VElement[name='template']" (rootNode) {
+        templateLocation = templateLocation || rootNode.loc.start
```
Comment (ota-meshi) [inline lib/rules/no-unregistered-components.js:104]: This variable is no longer needed.

File: lib/rules/no-unregistered-components.js (hunk 10 above)
Comment (ota-meshi) [reply lib/rules/no-unregistered-components.js:120]: This case is not valid.
Components registered with kebab-case cannot be used with PascalCase.

File: lib/rules/no-unregistered-components.js [hunk 13]
```diff
@@ -0,0 +1,134 @@
...
+          !node.value.expression
+        ) return
+
+        if (node.value.expression.type === 'Literal') {
+          if (utils.isHtmlWellKnownElementName(node.value.expression.value)) return
+          usedComponentNodes.push({ node, name: node.value.expression.value })
+        }
+      },
+      "VAttribute[directive=false][key.name='is']" (node) {
+        if (utils.isHtmlWellKnownElementName(node.value.value)) return
```
Comment (ota-meshi) [inline lib/rules/no-unregistered-components.js:100]: `node.value` may be null. (I just noticed this.)

e.g.

```vue
<component is />
```

Could you add this test case and change the source?

https://github.com/mysticatea/vue-eslint-parser/blob/master/docs/ast.md#vattribute

**Comment by jesusgn90** [comment]
@ota-meshi great insights into this, thanks for it. I've just pushed https://github.com/vuejs/eslint-plugin-vue/pull/1114/commits/7c3198c27c5bd76af1694c740def40c1b8a9c6e8 fixing (hopefully) all the new concerns.

File: lib/rules/no-unregistered-components.js (hunk 12 above)
Comment (ota-meshi) [reply lib/rules/no-unregistered-components.js:107]: Please remove the templateLocation variable.

**Comment by jesusgn90** [comment]
> Thank you for making the changes!
> 
> However, there is one request remaining.

Yep, code updated, I forgot that one :+1:

**Comment by yoyo837** [comment]
Hey guys, How to lint `Layout not found`:

```vue
<script setup>
// import Layout from '@/components/common/Layout.vue';
import { RouterView } from 'vue-router';
import '@/stores';
</script>

<template>
  <Layout>
    <RouterView />
  </Layout>
</template>
```

```js
    'vue/no-unregistered-components': [
      'error',
      {
        ignorePatterns: ['van(\\-\\w+)+', 'el(\\-\\w+)+'],
      },
    ],
```

**Pull request merged**